### PR TITLE
Add truncate context for audit logging

### PR DESCRIPTION
**Pull request by danielhlarkin**
### Scope & Purpose

Fix mistaken "internal" audit logging for truncate operation

- [x] :hankey: Bugfix 
- [ ] :pizza: New feature 
- [ ] :hammer: Refactoring 
- [ ] :book: CHANGELOG entry made
- [ ] :muscle: The behavior in this PR was *manually tested*
- [ ] :computer: The behavior change can be verified via automatic tests

#### Backports:

- [x] No backports required
- [ ] Backports required for:

Enterprise component: https://github.com/arangodb/enterprise/pull/549

Jenkins: https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/11823/ (blue except unrelated)

### CODE REVIEW

**Review by graetzer — CHANGES_REQUESTED** [review]
So was the options thing default constructed ? Where are the tests that catch this?

File: arangod/Cluster/ClusterMethods.cpp [hunk 1]
```diff
@@ -199,13 +199,14 @@ OperationResult handleResponsesFromAllShards(
     std::vector<futures::Try<arangodb::network::Response>>& responses,
     std::function<void(Result&, VPackBuilder&, ShardID&, VPackSlice)> handler,
     std::function<void(Result&, VPackBuilder&)> pre = [](Result&, VPackBuilder&) -> void {},
-    std::function<void(Result&, VPackBuilder&)> post = [](Result&, VPackBuilder&) -> void {}) {
+    std::function<void(Result&, VPackBuilder&)> post = [](Result&, VPackBuilder&) -> void {},
+    OperationOptions const options = OperationOptions()) {
```
Comment (jsteemann) [inline arangod/Cluster/ClusterMethods.cpp:203]: Is this intentionall passed by value? I also would prefer to not use default function parameters, as it makes it easy to overlook necessary changes at call sites.
```suggestion
    OperationOptions const& options) {
```

**Comment by jsteemann** [comment]
LGTM in general, but I also think we should have tests in place that verify the correct logging of the user name (vs. "internal") in case we are dealing with incoming HTTP/VST requests.

**Comment by graetzer** [comment]
please change the OptionsResult constructor so we do not miss passing the options in
```bash
diff --git a/arangod/Utils/OperationResult.h b/arangod/Utils/OperationResult.h
index a5e429beaae..78e619ecbdc 100644
--- a/arangod/Utils/OperationResult.h
+++ b/arangod/Utils/OperationResult.h
@@ -67,7 +67,7 @@ struct OperationResult final {
 
   // create result with details
   OperationResult(Result result, std::shared_ptr<VPackBuffer<uint8_t>> buffer,
-                  OperationOptions options = {},
+                  OperationOptions options,
                   std::unordered_map<int, size_t> countErrorCodes = std::unordered_map<int, size_t>())
       : result(std::move(result)),
         buffer(std::move(buffer)),
```

**Comment by danielhlarkin** [comment]
https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/11855/

File: arangod/Agency/Constituent.cpp [hunk 2]
```diff
@@ -140,28 +140,17 @@ void Constituent::termNoLock(term_t t, std::string const& votedFor) {
...
-      } catch (std::exception const& e) {
-        LOG_TOPIC("ac75f", FATAL, Logger::AGENCY)
-            << "Failed to replace  RAFT election ballot: " << e.what() << ". Bailing out.";
-        FATAL_ERROR_EXIT();
-      }
+    try {
+      OperationResult result = (tmp != t)
+                                   ? trx.insert("election", body.slice(), options)
+                                   : trx.replace("election", body.slice(), options);
+      res = trx.finish(result.errorNumber());
```
Comment (jsteemann) [inline arangod/Agency/Constituent.cpp:147]:
```suggestion
      res = trx.finish(result.result);
```

Comment (jsteemann) [reply arangod/Agency/Constituent.cpp:147]: Seems like there isn't (and never was) any code that would check `res` afterwards. 
An exception is caught here, but if there is no exception, `res` can still contain an error. :thinking:

File: arangod/Agency/State.cpp [hunk 3]
```diff
@@ -207,17 +204,16 @@ bool State::persistConf(index_t index, term_t term, uint64_t millis,
     THROW_ARANGO_EXCEPTION(res);
   }
 
-  OperationResult logResult, confResult;
   try {
-    logResult = trx.insert("log", log.slice(), _options);
-    confResult = trx.replace("configuration", configuration.slice(), _options);
+    OperationResult logResult = trx.insert("log", log.slice(), _options);
```
Comment (jsteemann) [inline arangod/Agency/State.cpp:208]: There isn't (and never was) any check for the value of `logResult`. So it may go wrong and this will not be noticed.

Comment (jsteemann) [reply arangod/Agency/State.cpp:208]: I guess we should bail out in that case as well.

File: arangod/Aql/SingleRemoteModificationExecutor.cpp [hunk 4]
```diff
@@ -107,13 +107,14 @@ template <typename Modifier>
 template <typename Modifier>
 auto SingleRemoteModificationExecutor<Modifier>::doSingleRemoteModificationOperation(
     InputAqlItemRow& input, Stats& stats) -> OperationResult {
-  OperationResult result;
   OperationOptions& options = _info._options;
```
Comment (jsteemann) [inline arangod/Aql/SingleRemoteModificationExecutor.cpp:110]: is the `options` variable needed here at all? it seems that the following code refers to `_info._options` instead.

File: arangod/Graph/GraphOperations.cpp [hunk 5]
```diff
@@ -170,90 +172,93 @@ OperationResult GraphOperations::eraseEdgeDefinition(bool waitForSync,
...
       GraphManager::getCollectionByName(_vocbase, vertexCollectionName);
 
   if (def == nullptr) {
-    return OperationResult(
-        Result(TRI_ERROR_GRAPH_VERTEX_COL_DOES_NOT_EXIST,
-               vertexCollectionName + " " +
-                   std::string{TRI_errno_string(TRI_ERROR_GRAPH_VERTEX_COL_DOES_NOT_EXIST)}));
+    return Result(Result(TRI_ERROR_GRAPH_VERTEX_COL_DOES_NOT_EXIST,
+                         vertexCollectionName + " " +
+                             std::string{TRI_errno_string(TRI_ERROR_GRAPH_VERTEX_COL_DOES_NOT_EXIST)}));
```
Comment (jsteemann) [inline arangod/Graph/GraphOperations.cpp:219]:
```suggestion
    return Result(TRI_ERROR_GRAPH_VERTEX_COL_DOES_NOT_EXIST,
                         vertexCollectionName + " " +
                             std::string{TRI_errno_string(TRI_ERROR_GRAPH_VERTEX_COL_DOES_NOT_EXIST)});
```

File: arangod/Agency/State.cpp [hunk 6]
```diff
@@ -207,17 +204,16 @@ bool State::persistConf(index_t index, term_t term, uint64_t millis,
...
+    OperationResult confResult =
+        trx.replace("configuration", configuration.slice(), _options);
+    res = trx.finish(confResult.result);
   } catch (std::exception const& e) {
     LOG_TOPIC("ced35", ERR, Logger::AGENCY) << "Failed to persist log entry:" << e.what();
     return false;
   }
```
Comment (jsteemann) [inline arangod/Agency/State.cpp:215]:
```suggestion
    OperationResult logResult = trx.insert("log", log.slice(), _options);
    if (logResult.fail()) {
      THROW_ARANGO_EXCEPTION(logResult.result);
    }
    OperationResult confResult =
        trx.replace("configuration", configuration.slice(), _options);
    res = trx.finish(confResult.result);
  } catch (std::exception const& e) {
    LOG_TOPIC("ced35", ERR, Logger::AGENCY) << "Failed to persist log entry: " << e.what();
    return false;
  }
```

**Comment by danielhlarkin** [comment]
https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/11886/

File: arangod/Agency/Constituent.cpp [hunk 7]
```diff
@@ -140,28 +140,17 @@ void Constituent::termNoLock(term_t t, std::string const& votedFor) {
...
-      } catch (std::exception const& e) {
-        LOG_TOPIC("ac75f", FATAL, Logger::AGENCY)
-            << "Failed to replace  RAFT election ballot: " << e.what() << ". Bailing out.";
-        FATAL_ERROR_EXIT();
-      }
+    try {
+      OperationResult result = (tmp != t)
+                                   ? trx.insert("election", body.slice(), options)
+                                   : trx.replace("election", body.slice(), options);
+      res = trx.finish(result.result);
```
Comment (graetzer) [inline arangod/Agency/Constituent.cpp:147]: ❤️

**Review by graetzer — APPROVED** [review]
well done!

**Pull request merged**